server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java

```diff
@@ -126,18 +131,27 @@ public static ShardRouting findNext(final ShardsIterator shardsIt, Exception fai
         return next;
     }
 
-    private static void logFailOpen(ShardId shardID) {
+    private void logFailOpen(ShardId shardID) {
         logger.info(() -> new ParameterizedMessage("{}: Fail open executed", shardID));
     }
 
     /**
      * *
-     * @param e exception causing failure for previous shard copy
      * @return true if can fail open ie request shard copies present in nodes with weighted shard
      * routing weight set to zero
      */
-    private static boolean canFailOpen(Exception e) {
-        return isInternalFailure(e);
+    private boolean canFailOpen(ShardId shardId) {
+        return isInternalFailure() || hasUnassignedShards(shardId);
+    }
+
+    private boolean hasUnassignedShards(ShardId shardId) {
+        List<ShardRouting> shards = clusterState.routingTable().shardRoutingTable(shardId).shards();
+        for (ShardRouting shardRouting : shards) {
+            if (shardRouting.unassigned()) {
```

Comment:
We should add `! shardRouting.active()` here, as it would include unassigned and initializing shards as well. And then rename method as `hasInActiveShardCopies` .